In [2]:
import math

import matplotlib.figure as figure
import matplotlib.pyplot as plt
import numpy as np
import numpy.matlib
import pandas as pd
import functions
from scipy.spatial.distance import cdist
from sklearn import metrics
from sklearn import svm
from sklearn.cross_decomposition import PLSRegression
from sklearn.model_selection import cross_val_predict, GridSearchCV
from sklearn.neighbors import NearestNeighbors

In [ ]:
# データセットの読み込み
dataset = pd.read_csv('debutanizer_y_10.csv', encoding='SHIFT-JIS', index_col=0)
y = dataset.iloc[:, 0:1]
x_tmp = dataset.iloc[:, 1:]

dataset

y     x1     x2     x3     x4     x5     x6     x7
sample_1     0.180  0.269  0.651  0.833  0.583  0.785  0.843  0.822
sample_2     0.180  0.268  0.650  0.852  0.578  0.776  0.839  0.822
sample_3     0.180  0.268  0.660  0.824  0.572  0.765  0.808  0.786
sample_4     0.180  0.267  0.668  0.808  0.566  0.753  0.800  0.786
sample_5     0.180  0.267  0.647  0.762  0.560  0.745  0.773  0.746
...            ...    ...    ...    ...    ...    ...    ...    ...
sample_2390  0.295  0.285  0.646  0.689  0.355  0.616  0.609  0.501
sample_2391  0.179  0.266  0.666  0.666  0.354  0.589  0.582  0.509
sample_2392  0.179  0.247  0.666  0.665  0.353  0.561  0.602  0.523
sample_2393  0.179  0.229  0.673  0.677  0.353  0.531  0.637  0.538
sample_2394  0.179  0.216  0.669  0.678  0.352  0.500  0.685  0.570

[2394 rows x 8 columns]

In [22]:
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# サブプロット作成
fig = make_subplots(rows=len(dataset.columns), cols=1, shared_xaxes=True, 
                    subplot_titles=dataset.columns, vertical_spacing=0.05)

# 各列に対して折れ線グラフを追加
for i, column in enumerate(dataset.columns):
    fig.add_trace(
        go.Scatter(x=dataset.index, y=dataset[column], mode="lines+markers", name=column),
        row=i + 1, col=1
    )

# レイアウト設定
fig.update_layout(
    height=300 * len(dataset.columns),  # グラフ全体の高さを動的に設定
    title="Line Plots for Each Column",
    showlegend=False  # 全体の凡例を非表示に（個別の凡例がサブプロットに表示されるため）
)

# 描画
fig.show()

In [ ]:
number_of_test_samples = 1000  # テストデータのサンプル数
method_name = 'jitsvr'  # ソフトセンサーの種類は以下の通りです
# 'pls' : 最初のトレーニングデータで構築した PLS モデル (適応的ソフトセンサーではありません)
# 'svr' : 最初のトレーニングデータで構築した SVR モデル (適応的ソフトセンサーではありません)
# 'mwpls' : Moving Window PLSs
# 'mwsvr' : Moving Window SVR
# 'jitpls' : Just-In-Time PLS
# 'jitsvr' : Just-In-Time SVR
# 'lwpls' : Locally-Weighted PLS

number_of_samples_in_modeling = 100  # MW や JIT モデルにおけるモデル構築用サンプルの数 
y_measurement_delay = 10  # y の測定時間の遅れ
dynamics_max = 10  # いくつまで時間遅れ変数を追加するか。0 なら時間遅れ変数は追加されません
dynamics_span = 2  # いくつずつ時間を遅らせた変数を追加するか
add_nonlinear_terms_flag = False  # True (二乗項・交差項を追加) or False (追加しない)

max_sample_size = 10000  # データベースにおける最大のサンプル数
fold_number = 5  # N-fold CV の N
max_number_of_principal_components = 20  # 使用する主成分の最大数
svr_cs = 2 ** np.arange(-5, 11, dtype=float)  # C の候補
svr_epsilons = 2 ** np.arange(-10, 1, dtype=float)  # ε の候補
svr_gammas = 2 ** np.arange(-20, 11, dtype=float)  # γ の候補
lwpls_lambdas = 2 ** np.arange(-9, 6, dtype=float)

In [57]:
# 説明変数の非線形変換
if add_nonlinear_terms_flag:
    x_tmp = functions.add_nonlinear_terms(x_tmp)  # 説明変数の二乗項や交差項を追加
    x = x_tmp.drop(x_tmp.columns[x_tmp.std() == 0], axis=1)  # 標準偏差が 0 の説明変数を削除
else:
    x = x_tmp.copy()
x

x1     x2     x3     x4     x5     x6     x7
sample_1     0.269  0.651  0.833  0.583  0.785  0.843  0.822
sample_2     0.268  0.650  0.852  0.578  0.776  0.839  0.822
sample_3     0.268  0.660  0.824  0.572  0.765  0.808  0.786
sample_4     0.267  0.668  0.808  0.566  0.753  0.800  0.786
sample_5     0.267  0.647  0.762  0.560  0.745  0.773  0.746
...            ...    ...    ...    ...    ...    ...    ...
sample_2390  0.285  0.646  0.689  0.355  0.616  0.609  0.501
sample_2391  0.266  0.666  0.666  0.354  0.589  0.582  0.509
sample_2392  0.247  0.666  0.665  0.353  0.561  0.602  0.523
sample_2393  0.229  0.673  0.677  0.353  0.531  0.637  0.538
sample_2394  0.216  0.669  0.678  0.352  0.500  0.685  0.570

[2394 rows x 7 columns]

- flag = False だと、そのままコピー

In [58]:
# 時間遅れ変数の追加
dataset_with_dynamics = functions.add_time_delayed_variable(pd.concat([y, x], axis=1), dynamics_max, dynamics_span)

dataset_with_dynamics

array([[0.155, 0.242, 0.263, ..., 0.746, 0.786, 0.822],
       [0.155, 0.231, 0.253, ..., 0.786, 0.786, 0.822],
       [0.155, 0.22 , 0.242, ..., 0.821, 0.746, 0.786],
       ...,
       [0.179, 0.247, 0.285, ..., 0.587, 0.627, 0.665],
       [0.179, 0.229, 0.266, ..., 0.555, 0.574, 0.642],
       [0.179, 0.216, 0.247, ..., 0.554, 0.587, 0.627]])

In [59]:
# トレーニングデータとテストデータに分割
dataset_train_with_dynamics = dataset_with_dynamics[:-number_of_test_samples, :]
dataset_test_with_dynamics = dataset_with_dynamics[-number_of_test_samples:, :]

# トレーニングデータでは y が測定されたサンプルのみ収集
y_measured_dataset_train = dataset_train_with_dynamics[0:1, :]
for sample_number in range(1, dataset_train_with_dynamics.shape[0]):
    if y_measured_dataset_train[-1, 0] != dataset_train_with_dynamics[sample_number, 0]:
        y_measured_dataset_train = np.r_[
            y_measured_dataset_train, dataset_train_with_dynamics[sample_number:sample_number + 1, :]]
y_measured_dataset_train = np.delete(y_measured_dataset_train, 0, 0)
y_measured_dataset_train = pd.DataFrame(y_measured_dataset_train)
y_train = y_measured_dataset_train.iloc[:, 0]
x_train = y_measured_dataset_train.iloc[:, 1:]

y_measured_dataset_train

0       1      2       3      4      5       6      7      8      9   \
0    0.1960  0.2340  0.162  0.2080  0.201  0.220  0.2420  0.669  0.648  0.647   
1    0.2490  0.0728  0.243  0.2410  0.239  0.236  0.2340  0.643  0.682  0.658   
2    0.2810  0.1740  0.139  0.0554  0.131  0.140  0.0728  0.654  0.670  0.638   
3    0.3090  0.1040  0.102  0.1010  0.203  0.254  0.1740  0.682  0.672  0.688   
4    0.3760  0.1510  0.139  0.1280  0.116  0.107  0.1040  0.663  0.638  0.684   
..      ...     ...    ...     ...    ...    ...     ...    ...    ...    ...   
133  0.0202  0.2310  0.235  0.3650  0.526  0.631  0.7360  0.638  0.667  0.647   
134  0.2530  0.2770  0.262  0.2460  0.231  0.227  0.2310  0.665  0.657  0.695   
135  0.4910  0.3720  0.348  0.3250  0.308  0.293  0.2770  0.697  0.689  0.667   
136  0.4190  0.3390  0.356  0.3730  0.390  0.396  0.3720  0.661  0.676  0.689   
137  0.3490  0.3710  0.358  0.3460  0.334  0.322  0.3390  0.709  0.681  0.666   

     ...     33     34     35     36     37     38     39     40     41     42  
0    ...  0.812  0.749  0.728  0.801  0.766  0.772  0.798  0.718  0.720  0.789  
1    ...  0.758  0.780  0.770  0.766  0.793  0.775  0.763  0.769  0.759  0.766  
2    ...  0.744  0.813  0.826  0.780  0.763  0.774  0.732  0.813  0.808  0.793  
3    ...  0.766  0.855  0.809  0.771  0.737  0.739  0.759  0.863  0.814  0.763  
4    ...  0.780  0.799  0.737  0.735  0.780  0.781  0.740  0.821  0.762  0.737  
..   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  
133  ...  0.221  0.485  0.520  0.574  0.206  0.134  0.206  0.450  0.462  0.508  
134  ...  0.410  0.438  0.335  0.239  0.403  0.409  0.372  0.420  0.292  0.206  
135  ...  0.449  0.435  0.422  0.431  0.492  0.414  0.414  0.409  0.373  0.403  
136  ...  0.417  0.437  0.534  0.530  0.436  0.434  0.362  0.402  0.479  0.492  
137  ...  0.423  0.445  0.444  0.444  0.459  0.394  0.398  0.426  0.415  0.436  

[138 rows x 43 columns]

In [60]:
# テストデータ
dataset_test_with_dynamics = pd.DataFrame(dataset_test_with_dynamics)
y_test_all = dataset_test_with_dynamics.iloc[:, 0]
x_test = dataset_test_with_dynamics.iloc[:, 1:]

In [61]:
# サンプルの調整
if method_name[0:2] == 'mw':
    y_train = y_train.iloc[-number_of_samples_in_modeling:]
    x_train = x_train.iloc[-number_of_samples_in_modeling:, :]
else:
    y_train = y_train.iloc[-max_sample_size:]
    x_train = x_train.iloc[-max_sample_size:, :]
    if method_name[0:3] == 'jit':
        nn_model = NearestNeighbors(metric='euclidean')  # サンプル選択用の k-NN モデルの宣言

In [62]:
# オートスケーリング    
autoscaled_x_train = (x_train - x_train.mean()) / x_train.std()
autoscaled_y_train = (y_train - y_train.mean()) / y_train.std()
autoscaled_x_test = (x_test - x_train.mean()) / x_train.std()

In [63]:
# ハイパーパラメータの最適化やモデリング
if method_name == 'pls' or method_name == 'mwpls':
    # CV による成分数の最適化
    components = []  # 空の list の変数を作成して、成分数をこの変数に追加していきます同じく成分数をこの変数に追加
    r2_in_cv_all = []  # 空の list の変数を作成して、成分数ごとのクロスバリデーション後の r2 をこの変数に追加
    for component in range(1, min(np.linalg.matrix_rank(autoscaled_x_train), max_number_of_principal_components) + 1):
        # PLS
        model = PLSRegression(n_components=component)  # PLS モデルの宣言
        estimated_y_in_cv = pd.DataFrame(cross_val_predict(model, autoscaled_x_train, autoscaled_y_train,
                                                           cv=fold_number))  # クロスバリデーション推定値の計算し、DataFrame型に変換
        estimated_y_in_cv = estimated_y_in_cv * y_train.std() + y_train.mean()  # スケールをもとに戻す
        r2_in_cv = metrics.r2_score(y_train, estimated_y_in_cv)  # r2 を計算
        print(component, r2_in_cv)  # 成分数と r2 を表示
        r2_in_cv_all.append(r2_in_cv)  # r2 を追加
        components.append(component)  # 成分数を追加
    optimal_component_number = components[r2_in_cv_all.index(max(r2_in_cv_all))]  # 最適成分数
    # PLS
    model = PLSRegression(n_components=optimal_component_number)  # モデルの宣言
    model.fit(autoscaled_x_train, autoscaled_y_train)  # モデルの構築
elif method_name == 'svr' or method_name == 'mwsvr' or method_name == 'jitsvr':
    # グラム行列の分散を最大化することによる γ の最適化
    variance_of_gram_matrix = list()
    for svr_gamma in svr_gammas:
        gram_matrix = np.exp(
            -svr_gamma * cdist(autoscaled_x_train, autoscaled_x_train, metric='seuclidean'))
        variance_of_gram_matrix.append(gram_matrix.var(ddof=1))
    optimal_svr_gamma = svr_gammas[np.where(variance_of_gram_matrix == np.max(variance_of_gram_matrix))[0][0]]
    # CV による ε の最適化
    model_in_cv = GridSearchCV(svm.SVR(kernel='rbf', C=3, gamma=optimal_svr_gamma), {'epsilon': svr_epsilons},
                               cv=fold_number, iid=False)
    model_in_cv.fit(autoscaled_x_train, autoscaled_y_train)
    optimal_svr_epsilon = model_in_cv.best_params_['epsilon']
    # CV による C の最適化
    model_in_cv = GridSearchCV(svm.SVR(kernel='rbf', epsilon=optimal_svr_epsilon, gamma=optimal_svr_gamma),
                               {'C': svr_cs}, cv=fold_number, iid=False)
    model_in_cv.fit(autoscaled_x_train, autoscaled_y_train)
    optimal_svr_c = model_in_cv.best_params_['C']
    # CV による γ の最適化
    model_in_cv = GridSearchCV(svm.SVR(kernel='rbf', epsilon=optimal_svr_epsilon, C=optimal_svr_c),
                               {'gamma': svr_gammas}, cv=fold_number, iid=False)
    model_in_cv.fit(autoscaled_x_train, autoscaled_y_train)
    optimal_svr_gamma = model_in_cv.best_params_['gamma']
    # 最適化された C, ε, γ
    print('C : {0}\nε : {1}\nGamma : {2}'.format(optimal_svr_c, optimal_svr_epsilon, optimal_svr_gamma))
    # SVR
    model = svm.SVR(kernel='rbf', C=optimal_svr_c, epsilon=optimal_svr_epsilon, gamma=optimal_svr_gamma)  # モデルの宣言
    model.fit(autoscaled_x_train, autoscaled_y_train)  # モデルの構築
elif method_name == 'lwpls':
    # CV によるハイパーパラメータの最適化
    autoscaled_x_train = np.array(autoscaled_x_train)
    autoscaled_y_train = np.array(autoscaled_y_train)
    y_train_array = np.array(y_train)
    r2cvs = np.empty(
        (min(np.linalg.matrix_rank(autoscaled_x_train), max_number_of_principal_components), len(lwpls_lambdas)))
    min_number = math.floor(x_train.shape[0] / fold_number)
    mod_numbers = x_train.shape[0] - min_number * fold_number
    index = np.matlib.repmat(np.arange(1, fold_number + 1, 1), 1, min_number).ravel()
    if mod_numbers != 0:
        index = np.r_[index, np.arange(1, mod_numbers + 1, 1)]
    indexes_for_division_in_cv = np.random.permutation(index)
    np.random.seed()
    for parameter_number, lambda_in_similarity in enumerate(lwpls_lambdas):
        estimated_y_in_cv = np.empty((len(y_train_array), r2cvs.shape[0]))
        for fold in np.arange(1, fold_number + 1, 1):
            autoscaled_x_train_in_cv = autoscaled_x_train[indexes_for_division_in_cv != fold, :]
            autoscaled_y_train_in_cv = autoscaled_y_train[indexes_for_division_in_cv != fold]
            autoscaled_x_validation_in_cv = autoscaled_x_train[indexes_for_division_in_cv == fold, :]

            estimated_y_validation_in_cv = functions.lwpls(autoscaled_x_train_in_cv, autoscaled_y_train_in_cv,
                                                                  autoscaled_x_validation_in_cv, r2cvs.shape[0],
                                                                  lambda_in_similarity)
            estimated_y_in_cv[indexes_for_division_in_cv == fold, :] = estimated_y_validation_in_cv * y_train_array.std(
                ddof=1) + y_train_array.mean()

        estimated_y_in_cv[np.isnan(estimated_y_in_cv)] = 99999
        ss = (y_train_array - y_train_array.mean()).T.dot(y_train_array - y_train_array.mean())
        press = np.diag(
            (np.matlib.repmat(y_train_array.reshape(len(y_train_array), 1), 1,
                              estimated_y_in_cv.shape[1]) - estimated_y_in_cv).T.dot(
                np.matlib.repmat(y_train_array.reshape(len(y_train_array), 1), 1,
                                 estimated_y_in_cv.shape[1]) - estimated_y_in_cv))
        r2cvs[:, parameter_number] = 1 - press / ss

    best_candidate_number = np.where(r2cvs == r2cvs.max())

    optimal_component_number = best_candidate_number[0][0] + 1
    optimal_lambda_in_similarity = lwpls_lambdas[best_candidate_number[1][0]]

TypeError: __init__() got an unexpected keyword argument 'iid'

- 各成分数によるR2の数値が出力
- 一番数値が高い成分数を選択する必要あり

In [ ]:
# y の推定やモデリング
if method_name == 'pls' or method_name == 'svr':
    estimated_y_test_all = model.predict(autoscaled_x_test) * y_train.std() + y_train.mean()
else:
    estimated_y_test_all = np.zeros((len(y_test_all)))
    for test_sample_number in range(len(y_test_all)):
        print(test_sample_number + 1, '/', len(y_test_all))
        autoscaled_x_test = (x_test.iloc[
                             test_sample_number:test_sample_number + 1, ] - x_train.mean()) / x_train.std()  # オートスケーリング
        # y の推定
        if method_name[0:2] == 'mw':
            autoscaled_estimated_y_test_tmp = model.predict(autoscaled_x_test)
        elif method_name == 'lwpls':
            autoscaled_estimated_y_test_tmp = functions.lwpls(autoscaled_x_train, autoscaled_y_train,
                                                                     autoscaled_x_test, optimal_component_number,
                                                                     optimal_lambda_in_similarity)
            autoscaled_estimated_y_test_tmp = autoscaled_estimated_y_test_tmp[:, optimal_component_number - 1]
        elif method_name[0:3] == 'jit':
            # サンプル選択
            nn_model.fit(autoscaled_x_train)
            tmp, nn_index_test = nn_model.kneighbors(autoscaled_x_test,
                                                     n_neighbors=min(number_of_samples_in_modeling, x_train.shape[0]))
            x_train_jit = x_train.iloc[nn_index_test[0, :], :]
            y_train_jit = y_train.iloc[nn_index_test[0, :]]
            # オートスケーリング    
            autoscaled_x_train_jit = (x_train_jit - x_train_jit.mean()) / x_train_jit.std()
            autoscaled_y_train_jit = (y_train_jit - y_train_jit.mean()) / y_train_jit.std()
            autoscaled_x_test_jit = (x_test.iloc[
                                     test_sample_number:test_sample_number + 1, ] - x_train_jit.mean()) / x_train_jit.std()
            # ハイパーパラメータの最適化
            if method_name == 'jitpls':
                # CV による成分数の最適化
                components = []  # 空の list の変数を作成して、成分数をこの変数に追加していきます同じく成分数をこの変数に追加
                r2_in_cv_all = []  # 空の list の変数を作成して、成分数ごとのクロスバリデーション後の r2 をこの変数に追加
                for component in range(1, min(np.linalg.matrix_rank(autoscaled_x_train_jit),
                                              max_number_of_principal_components) + 1):
                    model = PLSRegression(n_components=component)  # PLS モデルの宣言
                    estimated_y_in_cv = pd.DataFrame(
                        cross_val_predict(model, autoscaled_x_train_jit, autoscaled_y_train_jit,
                                          cv=fold_number))  # クロスバリデーション推定値の計算し、DataFrame型に変換
                    estimated_y_in_cv = estimated_y_in_cv * y_train_jit.std() + y_train_jit.mean()  # スケールをもとに戻す
                    r2_in_cv = metrics.r2_score(y_train_jit, estimated_y_in_cv)  # r2 を計算
                    r2_in_cv_all.append(r2_in_cv)  # r2 を追加
                    components.append(component)  # 成分数を追加
                optimal_component_number = components[r2_in_cv_all.index(max(r2_in_cv_all))]  # 最適成分数
                # PLS
                model = PLSRegression(n_components=optimal_component_number)  # モデルの宣言
            # モデリング
            model.fit(autoscaled_x_train_jit, autoscaled_y_train_jit)  # モデルの構築
            autoscaled_estimated_y_test_tmp = model.predict(autoscaled_x_test_jit)  # 推定
        if np.isnan(autoscaled_estimated_y_test_tmp):
            if test_sample_number == 0:
                estimated_y_test_all[test_sample_number] = 0
            else:
                estimated_y_test_all[test_sample_number] = estimated_y_test_all[test_sample_number - 1]
        else:
            estimated_y_test_all[test_sample_number] = autoscaled_estimated_y_test_tmp * y_train.std() + y_train.mean()

        if test_sample_number - y_measurement_delay >= 1:
            if y_test_all[test_sample_number - y_measurement_delay] - y_test_all[
                test_sample_number - y_measurement_delay - 1] != 0:
                x_train = pd.concat([x_train, x_test.iloc[
                                              test_sample_number - y_measurement_delay:test_sample_number - y_measurement_delay + 1, ]],
                                    axis=0)
                y_train = pd.concat([y_train, y_test_all.iloc[test_sample_number - y_measurement_delay:test_sample_number - y_measurement_delay + 1]])
                # サンプルの調整
                if method_name[0:2] == 'mw':
                    y_train = y_train.iloc[-number_of_samples_in_modeling:]
                    x_train = x_train.iloc[-number_of_samples_in_modeling:, :]
                else:
                    y_train = y_train.iloc[-max_sample_size:]
                    x_train = x_train.iloc[-max_sample_size:, :]
                # オートスケーリング    
                autoscaled_x_train = (x_train - x_train.mean()) / x_train.std()
                autoscaled_y_train = (y_train - y_train.mean()) / y_train.std()
                # ハイパーパラメータの最適化
                if method_name == 'mwpls':
                    # CV による成分数の最適化
                    components = []  # 空の list の変数を作成して、成分数をこの変数に追加していきます同じく成分数をこの変数に追加
                    r2_in_cv_all = []  # 空の list の変数を作成して、成分数ごとのクロスバリデーション後の r2 をこの変数に追加
                    for component in range(1, min(np.linalg.matrix_rank(autoscaled_x_train),
                                                  max_number_of_principal_components) + 1):
                        model = PLSRegression(n_components=component)  # PLS モデルの宣言
                        estimated_y_in_cv = pd.DataFrame(
                            cross_val_predict(model, autoscaled_x_train, autoscaled_y_train,
                                              cv=fold_number))  # クロスバリデーション推定値の計算し、DataFrame型に変換
                        estimated_y_in_cv = estimated_y_in_cv * y_train.std() + y_train.mean()  # スケールをもとに戻す
                        r2_in_cv = metrics.r2_score(y_train, estimated_y_in_cv)  # r2 を計算
                        r2_in_cv_all.append(r2_in_cv)  # r2 を追加
                        components.append(component)  # 成分数を追加
                    optimal_component_number = components[r2_in_cv_all.index(max(r2_in_cv_all))]  # 最適成分数
                    # PLS
                    model = PLSRegression(n_components=optimal_component_number)  # モデルの宣言
                # モデリング
                if method_name[0:2] == 'mw':
                    model.fit(autoscaled_x_train, autoscaled_y_train)  # モデルの構築

1 / 1000
2 / 1000
3 / 1000
4 / 1000
5 / 1000
6 / 1000
7 / 1000
8 / 1000
9 / 1000
10 / 1000
11 / 1000
12 / 1000
13 / 1000
14 / 1000
15 / 1000
16 / 1000
17 / 1000
18 / 1000
19 / 1000
20 / 1000
21 / 1000
22 / 1000
23 / 1000
24 / 1000
25 / 1000
26 / 1000
27 / 1000
28 / 1000
29 / 1000
30 / 1000
31 / 1000
32 / 1000
33 / 1000
34 / 1000
35 / 1000
36 / 1000
37 / 1000
38 / 1000
39 / 1000
40 / 1000
41 / 1000
42 / 1000
43 / 1000
44 / 1000
45 / 1000
46 / 1000
47 / 1000
48 / 1000
49 / 1000
50 / 1000
51 / 1000
52 / 1000
53 / 1000
54 / 1000
55 / 1000
56 / 1000
57 / 1000
58 / 1000
59 / 1000
60 / 1000
61 / 1000
62 / 1000
63 / 1000
64 / 1000
65 / 1000
66 / 1000
67 / 1000
68 / 1000
69 / 1000
70 / 1000
71 / 1000
72 / 1000
73 / 1000
74 / 1000
75 / 1000
76 / 1000
77 / 1000
78 / 1000
79 / 1000
80 / 1000
81 / 1000
82 / 1000
83 / 1000
84 / 1000
85 / 1000
86 / 1000
87 / 1000
88 / 1000
89 / 1000
90 / 1000
91 / 1000
92 / 1000
93 / 1000
94 / 1000
95 / 1000
96 / 1000
97 / 1000
98 / 1000
99 / 1000
100 / 1000
101 / 10

In [ ]:
estimated_y_test_all = pd.DataFrame(estimated_y_test_all, index=x_test.index, columns=['estimated_y'])
estimated_y_test_all.to_csv('estimated_y_test.csv')  # 推定値を csv ファイルに保存。同じ名前のファイルがあるときは上書きされますので注意してください

In [ ]:
# テストデータの y の実測値と推定値において y が測定されたサンプルのみ収集
ys_test = pd.concat([y_test_all, estimated_y_test_all], axis=1)
y_measured_ys_test = ys_test.iloc[0:1, :]
measured_index_test = []
for sample_number in range(1, ys_test.shape[0]):
    if y_measured_ys_test.iloc[-1, 0] != ys_test.iloc[sample_number, 0]:
        y_measured_ys_test = pd.concat([y_measured_ys_test, ys_test.iloc[sample_number:sample_number + 1, :]], axis=0)
        measured_index_test.append(sample_number)
y_measured_ys_test = y_measured_ys_test.drop(0, axis=0)

In [ ]:
# # テストデータの実測値 vs. 推定値のプロット
# plt.rcParams['font.size'] = 18  # 横軸や縦軸の名前の文字などのフォントのサイズ
# plt.figure(figsize=figure.figaspect(1))  # 図の形を正方形に
# plt.scatter(y_measured_ys_test.iloc[:, 0], y_measured_ys_test.iloc[:, 1], c='blue')  # 実測値 vs. 推定値プロット
# y_max = max(y_measured_ys_test.iloc[:, 0].max(), y_measured_ys_test.iloc[:, 1].max())  # 実測値の最大値と、推定値の最大値の中で、より大きい値を取得
# y_min = min(y_measured_ys_test.iloc[:, 0].min(), y_measured_ys_test.iloc[:, 1].min())  # 実測値の最小値と、推定値の最小値の中で、より小さい値を取得
# plt.plot([y_min - 0.05 * (y_max - y_min), y_max + 0.05 * (y_max - y_min)],
#          [y_min - 0.05 * (y_max - y_min), y_max + 0.05 * (y_max - y_min)], 'k-')  # 取得した最小値-5%から最大値+5%まで、対角線を作成
# plt.ylim(y_min - 0.05 * (y_max - y_min), y_max + 0.05 * (y_max - y_min))  # y 軸の範囲の設定
# plt.xlim(y_min - 0.05 * (y_max - y_min), y_max + 0.05 * (y_max - y_min))  # x 軸の範囲の設定
# plt.xlabel('actual y')  # x 軸の名前
# plt.ylabel('estimated y')  # y 軸の名前
# plt.show()  # 以上の設定で描画

In [ ]:
import plotly.graph_objects as go

# 実測値と推定値
actual_y = y_measured_ys_test.iloc[:, 0]
estimated_y = y_measured_ys_test.iloc[:, 1]

# 実測値と推定値の最大値・最小値を取得
y_max = max(actual_y.max(), estimated_y.max())
y_min = min(actual_y.min(), estimated_y.min())

# プロット範囲を調整（-5%～+5%）
margin = 0.05 * (y_max - y_min)
plot_min = y_min - margin
plot_max = y_max + margin

# Figureの作成
fig = go.Figure()

# 実測値 vs. 推定値の散布図
fig.add_trace(go.Scatter(
    x=actual_y,
    y=estimated_y,
    mode='markers',
    marker=dict(color='blue'),
    name='実測値 vs. 推定値'
))

# 対角線の追加
fig.add_trace(go.Scatter(
    x=[plot_min, plot_max],
    y=[plot_min, plot_max],
    mode='lines',
    line=dict(color='black', dash='solid'),
    name='対角線'
))

# レイアウトの設定
fig.update_layout(
    title='実測値 vs. 推定値',
    xaxis=dict(title='actual y', range=[plot_min, plot_max]),
    yaxis=dict(title='estimated y', range=[plot_min, plot_max]),
    width=600,  # 正方形に近い形状
    height=600,
    font=dict(size=18)
)

# 描画
fig.show()

In [ ]:
# テストデータのr2, RMSE, MAE
print('r^2 for test data :', metrics.r2_score(y_measured_ys_test.iloc[:, 0], y_measured_ys_test.iloc[:, 1]))
print('RMSE for test data :',
      metrics.mean_squared_error(y_measured_ys_test.iloc[:, 0], y_measured_ys_test.iloc[:, 1]) ** 0.5)
print('MAE for test data :', metrics.mean_absolute_error(y_measured_ys_test.iloc[:, 0], y_measured_ys_test.iloc[:, 1]))

r^2 for test data : 0.19821343274765435
RMSE for test data : 0.159297925478878
MAE for test data : 0.11072985113831021


- 'mwpls'
- r^2 for test data : 0.4327039984724812
- RMSE for test data : 0.13399407448530026
- MAE for test data : 0.09587478712379717

- lwpls
- r^2 for test data : 0.19821343274765435
- RMSE for test data : 0.159297925478878
- MAE for test data : 0.11072985113831021

In [ ]:
# # 実測値と推定値の時間プロット
# plt.rcParams['font.size'] = 18
# plt.plot(range(estimated_y_test_all.shape[0]), estimated_y_test_all.iloc[:, 0], 'b.-',
#          label='estimated y')
# plt.plot(measured_index_test, y_measured_ys_test.iloc[:, 0], 'r.', markersize=15, label='actual y')
# plt.xlabel('time')
# plt.ylabel(dataset.columns[0])
# plt.xlim([0, estimated_y_test_all.shape[0] + 1])
# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0, fontsize=18)
# plt.show()
# # 実測値と推定値の時間プロット (拡大)
# plt.plot(range(estimated_y_test_all.shape[0]), estimated_y_test_all.iloc[:, 0], 'b.-',
#          label='estimated y')
# plt.plot(measured_index_test, y_measured_ys_test.iloc[:, 0], 'r.', markersize=15, label='actual y')
# plt.xlabel('time')
# plt.ylabel(dataset.columns[0])
# plt.xlim([600, 800])
# plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0, fontsize=18)
# plt.show()

In [ ]:
import plotly.graph_objects as go

# 実測値と推定値
time = range(estimated_y_test_all.shape[0])  # 時間軸
estimated_y = estimated_y_test_all.iloc[:, 0]
actual_y_time = [time[i] for i in measured_index_test]
actual_y = y_measured_ys_test.iloc[:, 0]

# 図全体のプロット
fig_full = go.Figure()

# 推定値の時間プロット
fig_full.add_trace(go.Scatter(
    x=list(time),
    y=estimated_y,
    mode='lines+markers',
    name='estimated y',
    line=dict(color='blue'),
    marker=dict(color='blue')
))

# 実測値の時間プロット
fig_full.add_trace(go.Scatter(
    x=actual_y_time,
    y=actual_y,
    mode='markers',
    name='actual y',
    marker=dict(color='red', size=12)
))

# レイアウトの設定
fig_full.update_layout(
    title='Time Plot (Full Range)',
    xaxis=dict(title='time', range=[0, estimated_y_test_all.shape[0] + 1]),
    yaxis=dict(title=dataset.columns[0]),
    legend=dict(x=1.05, y=1, borderwidth=1, font=dict(size=18)),
    font=dict(size=18)
)

# 描画
fig_full.show()

# 拡大版のプロット
fig_zoomed = go.Figure(fig_full)  # 全体プロットをコピー
fig_zoomed.update_layout(
    title='Time Plot (Zoomed)',
    xaxis=dict(range=[600, 800])
)

# 描画
fig_zoomed.show()